### PR TITLE
ENH: add utility functions for repr printing

### PR DESCRIPTION
**Pull request by kohr-h**
This is a break-out from the monster PR #1276. It only adds the tools for better `repr` printing, thereby slightly changing how stuff works behind the scenes.
Follow-up PRs will apply this to the classes in ODL and fix some doc issues.

I've also started to implement #1311 here (more specific TODO comments).

### CODE REVIEW

**Comment by kohr-h** [comment]
Failure is due to Numpy 1.14.0 being installed, needs fix

File: odl/util/utility.py [hunk 1]
```diff
@@ -72,12 +78,80 @@ def indent(string, indent_str='    '):
...
+    <->This is line 1.
+    <->Next line.
+    <->And another one.
     """
-    return '\n'.join(indent_str + row for row in string.split('\n'))
+    return '\n'.join(indent_str + row for row in string.splitlines())
+
+
+def dedent(string, indent_str='   ', max_levels=None):
+    """Revert the effect of indentation.
```
Comment (adler-j) [inline odl/util/utility.py]: `IndentationOperator.inverse`!

(no, I'm not serious).

File: odl/util/utility.py [hunk 2]
```diff
@@ -72,12 +78,80 @@ def indent(string, indent_str='    '):
...
+    <->Level 2.
+    <-><->Level 3.
+    """
+    lines = string.splitlines()
+
+    # Determine common (minumum) number of indentation levels, capped at
+    # `max_levels` if given
+    def num_indents(line):
+        nindents = 0
+        while True:
```
Comment (adler-j) [inline odl/util/utility.py]: This is a scary risk for an infinite loop. Perhaps some safety?

File: odl/util/utility.py [hunk 3]
```diff
@@ -8,25 +8,31 @@
...
+__all__ = (
+    'array_str', 'dtype_str', 'dtype_repr', 'npy_printoptions',
+    'signature_string', 'signature_string_parts', 'repr_string',
+    'indent', 'dedent', 'attribute_repr_string', 'method_repr_string',
+    'is_numeric_dtype', 'is_int_dtype', 'is_floating_dtype', 'is_real_dtype',
+    'is_real_floating_dtype', 'is_complex_floating_dtype',
+    'real_dtype', 'complex_dtype', 'is_string', 'nd_iterator', 'conj_exponent',
+    'writable_array', 'run_from_ipython', 'NumpyRandomSeed',
+    'cache_arguments', 'unique',
+    'REPR_PRECISION')
```
Comment (adler-j) [inline odl/util/utility.py]: This file is getting out of hand. Perhaps we should split it (but leave that for another PR).

File: odl/util/utility.py [hunk 4]
```diff
@@ -72,12 +78,80 @@ def indent(string, indent_str='    '):
...
+    >>> print(text)
+    <->This is line 1.
+    <->Next line.
+    <->And another one.
+    >>> print(dedent(text, '<->'))
+    This is line 1.
+    Next line.
+    And another one.
+
+    Multiple levels of indentation:
```
Comment (adler-j) [inline odl/util/utility.py]: IMO this is too many tests, hard to read. At least add some spacing

File: odl/util/utility.py [hunk 5]
```diff
@@ -833,15 +995,436 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+        # Next string might fit on same line
+        separators.append(', ')
+        cur_line_len += 1  # for the extra space
+    else:
+        # Use linebreak if string contains newline or doesn't fit
+        separators.append(',\n')
+        cur_line_len = indent_len
+
+    for i, s in enumerate(strings[1:-1]):
+
```
Comment (adler-j) [inline odl/util/utility.py]: Remove blankline

File: odl/util/utility.py [hunk 6]
```diff
@@ -833,15 +995,436 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+    return tuple(separators)
+
+
+def repr_string(outer_string, inner_strings, allow_mixed_seps=True):
+    r"""Return a pretty string for ``repr``.
+
+    The returned string is formatted such that it does not extend
+    beyond the line boundary if avoidable. The line width is taken from
+    NumPy's printing options that can be retrieved with
+    ``np.get_printoptions()``. They can be temporarily overridden
```
Comment (adler-j) [inline odl/util/utility.py]: numpy.get_printoptions creates a clickable link

File: odl/util/utility.py [hunk 7]
```diff
@@ -833,15 +995,436 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+    using the `npy_printoptions` context manager. See Examples for details.
```
Comment (adler-j) [inline odl/util/utility.py]: This does not create a link

File: odl/util/utility.py (hunk 1 above)
Comment (kohr-h) [reply odl/util/utility.py]: :scream:

**Comment by kohr-h** [comment]
> But I feel that the code is somewhat abusing doctests to do full scale testing. I'd prefer doctests to stay small and perhaps move something to a real external test.

Well, here the intention is to document the behavior of the function as precisely as possible, and for that, unit tests are not a good vehicle. Doctests are much better suited for that. The functions here are not really user-facing, so I took the approach of maximizing documentation for developers.

File: odl/util/utility.py (hunk 7 above)
Comment (kohr-h) [reply odl/util/utility.py]: It should, since it's one of our functions in the same module

File: odl/util/utility.py (hunk 6 above)
Comment (kohr-h) [reply odl/util/utility.py]: ok

File: odl/util/utility.py (hunk 3 above)
Comment (kohr-h) [reply odl/util/utility.py]: See #1386

**Comment by kohr-h** [comment]
Comments addressed, rebased. Good to go?

**Comment by kohr-h** [comment]
Bump. Good to go in?
